File: kakao.app/js/ked/edit/tool/misc.js
var _k_ = {list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}}

var misc

import kxk from "../../../kxk.js"
let kutil = kxk.kutil
let kseg = kxk.kseg

import pepe from "../../../kxk/pepe.js"

import belt from "./belt.js"


misc = (function ()
{
    function misc ()
    {}

    misc["prepareWordsForCompletion"] = function (before, turd, words)
    {
        var balanced, beforeTurd, end, filtered, fix, fst, idx, lst, p, push, rwd, segl, segls, strs, subw, tc, w, wds

        filtered = []
        var list = _k_.list(words)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            w = list[_a_]
            if (w.startsWith(turd))
            {
                if (w.startsWith('..'))
                {
                    continue
                }
                if (w.startsWith('./'))
                {
                    continue
                }
                if (/\d+\.\d+/.test(w) && !w.startsWith(before))
                {
                    continue
                }
                p = pepe(w)
                if (p.unbalanced || p.mismatch)
                {
                    if (w.endsWith(p.tail) && p.mismatch)
                    {
                        fix = p.mismatch[0].content[0]
                        filtered.push(fix)
                        if (!_k_.empty((wds = kseg.words(fix))))
                        {
                            fst = fix.slice(0, wds[0].index + wds[0].word.length)
                            if (fst !== fix)
                            {
                                filtered.push(fst)
                            }
                        }
                    }
                    else if (p.unbalanced)
                    {
                        fix = p.unbalanced[0].content[0]
                        filtered.push(fix)
                    }
                    else
                    {
                    }
                    continue
                }
                if (0 <= w.indexOf('\n'))
                {
                    w = w.slice(0, typeof w.indexOf('\n') === 'number' ? w.indexOf('\n') : -1)
                }
                if (!_k_.empty((wds = kseg.words(w))))
                {
                    lst = wds[0].index + wds[0].word.length
                    if (lst < w.length - 1)
                    {
                        if (lst > before.length)
                        {
                            fst = w.slice(0, typeof lst === 'number' ? lst+1 : Infinity)
                            p = pepe(fst)
                            if (!p.unbalanced || p.mismatch)
                            {
                                filtered.push(fst)
                                continue
                            }
                            else
                            {
                                balanced = true
                            }
                        }
                        if (!balanced)
                        {
                            continue
                        }
                    }
                }
                filtered.push(w)
                continue
            }
            wds = kseg.words(w)
            var list1 = _k_.list(wds)
            for (idx = 0; idx < list1.length; idx++)
            {
                subw = list1[idx]
                if (subw.word.startsWith(turd))
                {
                    filtered.push(subw.word)
                    if (idx === wds.length - 1 && subw.index + subw.word.length < w.length - 1)
                    {
                        rwd = w.slice(subw.index)
                        p = pepe(rwd)
                        if (!(p.unbalanced || p.mismatch))
                        {
                            filtered.push(rwd)
                        }
                    }
                }
                else if (turd.length === 1 && turd === w[subw.index - 1])
                {
                    if (turd !== '.')
                    {
                        filtered.push(turd.slice(-1)[0] + subw.word)
                    }
                    else
                    {
                    }
                    if (/\d+\./.test(before) && /\d+/.test(subw.word))
                    {
                        if (w.startsWith(before))
                        {
                            filtered.push(turd.slice(-1)[0] + subw.word)
                        }
                    }
                    else if (!/\d+\./.test(before) && !/\d+/.test(subw.word))
                    {
                        filtered.push(turd.slice(-1)[0] + subw.word)
                    }
                }
            }
        }
        words = kutil.uniq(filtered)
        if (_k_.empty(words))
        {
            return []
        }
        segls = []
        push = function (s)
        {
            var ws

            if (kseg.str(s) === turd)
            {
                return
            }
            ws = kseg.words(s)
            ws = ws.filter(function (w)
            {
                return w.index + w.segl.length > turd.length
            })
            if (!_k_.empty(ws[0]))
            {
                if (ws[0].index === 0 && (turd !== ws[0].word && ws[0].word !== s) && ws[0].word.startsWith(turd))
                {
                    segls.push(ws[0].segl)
                }
            }
            return segls.push(s)
        }
        var list2 = _k_.list(kseg.segls(words))
        for (var _c_ = 0; _c_ < list2.length; _c_++)
        {
            segl = list2[_c_]
            tc = kseg.tailCountTurd(segl)
            if (tc === 0 || tc === 1 && segl[0] === segl.slice(-1)[0])
            {
                push(segl)
            }
            else
            {
                end = kseg.str(segl.slice(segl.length - tc))
                if (_k_.in(end,'])}') || _k_.in(end.slice(-1)[0],')]}') && !(_k_.in(end.slice(-2,-1)[0],',')))
                {
                    push(segl)
                }
                else
                {
                    beforeTurd = segl.slice(0, segl.length - tc)
                    if (!_k_.empty(beforeTurd))
                    {
                        push(segl.slice(0, segl.length - tc))
                    }
                }
            }
        }
        strs = segls.map(kseg.str)
        strs.sort()
        return kutil.uniq(strs)
    }

    misc["isCommentLine"] = function (line)
    {
        var trimmed

        trimmed = kseg.trim(line)
        return trimmed[0] === "#"
    }

    misc["indexOfExtremePositionInDirection"] = function (posl, dir, index)
    {
        var ci, cmp, exi, idx, pos, start

        switch (dir)
        {
            case 'left':
            case 'right':
                ci = 0
                break
            case 'up':
            case 'down':
                ci = 1
                break
        }

        switch (dir)
        {
            case 'left':
            case 'up':
                start = Infinity
                cmp = function (a, b)
                {
                    return a > b
                }
                break
            case 'right':
            case 'down':
                start = -Infinity
                cmp = function (a, b)
                {
                    return a < b
                }
                break
        }

        var list = _k_.list(posl)
        for (idx = 0; idx < list.length; idx++)
        {
            pos = list[idx]
            if (cmp(start,pos[ci]))
            {
                start = pos[ci]
                exi = idx
            }
        }
        if (!_k_.empty(index) && !_k_.empty(posl[index]))
        {
            if (posl[exi][ci] === posl[index][ci])
            {
                return index
            }
        }
        return exi
    }

    return misc
})()

export default misc;

File: kakao.app/love/lua/edit/tool/misc.lua
--[[
    00     00  000   0000000   0000000  
    000   000  000  000       000       
    000000000  000  0000000   000       
    000 0 000  000       000  000       
    000   000  000  0000000    0000000  
--]]

-- use ../../../kxk ▪ kutil kseg
-- use ../../../kxk ◆ pepe
-- use              ◆ belt


local misc = class("misc")
    

--[[
     0000000   0000000   00     00  00000000   000      00000000  000000000  000   0000000   000   000  
    000       000   000  000   000  000   000  000      000          000     000  000   000  0000  000  
    000       000   000  000000000  00000000   000      0000000      000     000  000   000  000 0 000  
    000       000   000  000 0 000  000        000      000          000     000  000   000  000  0000  
     0000000   0000000   000   000  000        0000000  00000000     000     000   0000000   000   000  

    args:
        before   the chunk before the cursor
        turd     the part of it that needs completion, usually the last word or the last punctuation character  
        words    list of chunks collected from current file and other sources
    
    ⮐  list of strings suitable for completion
    
    filters chunks that don't match
    removes trailing crappy punctuation from the matching chunks
    inserts matching subwords if chunk contains punctuation, eg: for chunk 'obj.prop' and before 'my.' it will include '.prop'
    --]]


function misc.static.prepareWordsForCompletion(before, turd, words) 
        local filtered = array()
        
        for wi, w in ipairs(words) do 
            if kstr.startsWith(w, turd) then 
                if kstr.startsWith(w, '..') then _ = nil
                elseif kstr.startsWith(w, './') then _ = nil
                elseif (string.match(w, "\\d+.\\d+") and not w.startsWith(before)) then _ = nil
                else 
                        local p = pepe(w)
                        if (p.unbalanced or p.mismatch) then 
                            if (w.endsWith(p.tail) and p.mismatch) then 
                                local fix = p.mismatch[1].content[1]
                                filtered:push(fix)
                                local wds = kseg.words(fix)
                                if valid(wds) then 
                                    local fst = string.sub(fix, 1, (wds[1].index + #wds[1].word))
                                    if (fst ~= fix) then 
                                        filtered:push(fst)
                                    end
                                end
                            elseif p.unbalanced then 
                                local fix = p.unbalanced[1].content[1]
                                filtered:push(fix)
                            end
                        else 
                            if (0 <= kstr.find(w, '\n')) then 
                                w = string.sub(w, 1, kstr.find(w, '\n'))
                            end
                            
                            local push = true
                            local wds = kseg.words(w)
                            if valid(wds) then 
                                local balanced = false
                                local lst = (wds[1].index + #wds[1].word)
                                if (lst < #w) then 
                                    if (lst >= #before) then 
                                        local fst = string.sub(w, 1, lst)
                                        p = pepe(fst)
                                        if (not p.unbalanced or p.mismatch) then 
                                            filtered:push(fst)
                                            push = false
                                        else 
                                            balanced = true
                                        end
                                    end
                                    
                                    if not balanced then 
                                        push = false
                                    end
                                end
                            end
                            
                            if push then 
                                filtered:push(w)
                            end
                        end
                end
            else 
                local wds = kseg.words(w)
                for idx, subw in ipairs(wds) do 
                    if subw.word.startsWith(turd) then 
                        filtered:push(subw.word)
                        
                        if ((idx == (#wds - 1)) and ((subw.index + #subw.word) < (#w - 1))) then 
                            local rwd = string.sub(w, subw.index)
                            local p = pepe(rwd)
                            if not (p.unbalanced or p.mismatch) then 
                                filtered:push(rwd)
                            end
                        end
                    elseif ((#turd == 1) and (turd == w[(subw.index - 1)])) then 
                        if (turd ~= '.') then 
                            filtered:push((turd[-1] + subw.word))
                        else 
                            if (string.match(before, "\\d+\\.") and string.match(subw.word, "\\d+")) then 
                                if w.startsWith(before) then 
                                    filtered:push((turd[-1] + subw.word))
                                end
                            elseif (not string.match(before, "\\d+\\.") and not string.match(subw.word, "\\d+")) then 
                                filtered:push((turd[-1] + subw.word))
                            end
                        end
                    end
                end
            end
        end
        
        words = kutil.uniq(filtered)
        
        if empty(words) then return array() end
        
        local segls = array()
        
        
        function push(s) 
            if (kseg.str(s) == turd) then return end
            
            local ws = kseg.words(s)
            ws = ws.filter(function (w) 
    return ((w.index + #w.segl) > #turd)
end)
            
            if valid(ws[0]) then 
                if (((ws[0].index == 0) and ((turd ~= ws[0].word) ~= s)) and ws[0].word.startsWith(turd)) then 
                    segls.push(ws[0].segl)
                end
            end
            
            return segls.push(s)
        end
        
        for si, segl in ipairs(kseg.segls(words)) do 
            local tc = kseg.tailCountTurd(segl)
            
            if ((tc == 0) or ((tc == 1) and (segl[0] == segl[-1]))) then 
                push(segl)
            else 
                local ends = kseg.str(segl:slice((#segl - tc)))
                if ((kstr.find('])}', ends) >= 1) or ((kstr.find('])}', string.sub(ends, #ends, #ends)) >= 1) and (string.sub(ends, (#ends - 1), (#ends - 1)) ~= ','))) then 
                    push(segl)
                else 
                    local beforeTurd = segl:slice(1, (#segl - tc))
                    if valid(beforeTurd) then 
                        push(segl:slice(1, (#segl - tc)))
                    end
                end
            end
        end
        
        local strs = segls.map(kseg.str)
        strs.sort()
        return kutil.uniq(strs)
    end

--  0000000   0000000   00     00  00     00  00000000  000   000  000000000  
-- 000       000   000  000   000  000   000  000       0000  000     000     
-- 000       000   000  000000000  000000000  0000000   000 0 000     000     
-- 000       000   000  000 0 000  000 0 000  000       000  0000     000     
--  0000000   0000000   000   000  000   000  00000000  000   000     000     


function misc.static.isCommentLine(line) 
        local trimmed = kseg.trim(line)
        return (trimmed[0] == "#")
    end


function misc.static.indexOfExtremePositionInDirection(posl, dir, index) 
        local cmp = nil
        local exi = nil
        local start = nil
        local ci = 1
        
        if (dir == 'left') or (dir == 'right') then ci = 1
        elseif (dir == 'up') or (dir == 'down') then ci = 2
        end
        
        if (dir == 'left') or (dir == 'up') then start = Infinity ; function cmp(a, b) 
    return (a > b)
                                                 end
        elseif (dir == 'right') or (dir == 'down') then start = -Infinity ; function cmp(a, b) 
    return (a < b)
                                                 end
        end
        
        for idx, pos in ipairs(posl) do 
            if cmp(start, pos[ci]) then 
                start = pos[ci]
                exi = idx
            end
        end
        
        if (valid(index) and valid(posl[index])) then 
            if (posl[exi][ci] == posl[index][ci]) then 
                return index
            end
        end
        
        return exi
    end

return misc